empyrical/perf_attrib.py

```diff
@@ -0,0 +1,409 @@
+            dt         ticker
+            2017-01-01 AAPL   -1.592914  0.852830
+                       TLT     0.184864  0.895534
+                       XOM     0.993160  1.149353
+            2017-01-02 AAPL   -0.140009 -0.524952
+                       TLT    -1.066978  0.185435
+                       XOM    -1.798401  0.761549
+
+    stack_positions : bool
+        Flag indicating whether `positions` should be converted to long format.
+
+    pos_in_dollars : bool
+        Flag indicating whether `positions` are in dollars or percentages
+        If True, positions are in dollars.
+
+    Returns
+    -------
+    risk_exposures_portfolio : pd.DataFrame
+        df indexed by datetime, with factors as columns
+        - Example:
+                        momentum  reversal
+            dt
+            2017-01-01 -0.238655  0.077123
+            2017-01-02  0.821872  1.520515
+    """
+    if stack_positions:
+        positions = _stack_positions(positions, pos_in_dollars=pos_in_dollars)
+
+    risk_exposures = factor_loadings.multiply(positions,
+                                              axis='rows')
+
+    return risk_exposures.groupby(level='dt').sum()
+
+
+def create_perf_attrib_stats(perf_attrib, risk_exposures):
+    """
+    Takes perf attribution data over a period of time and computes annualized
+    multifactor alpha, multifactor sharpe, risk exposures.
+    """
+    summary = OrderedDict()
+    total_returns = perf_attrib['total_returns']
+    specific_returns = perf_attrib['specific_returns']
+    common_returns = perf_attrib['common_returns']
+
+    summary['Annualized Specific Return'] =\
+        annual_return(specific_returns)
+    summary['Annualized Common Return'] =\
+        annual_return(common_returns)
+    summary['Annualized Total Return'] =\
+        annual_return(total_returns)
+
+    summary['Specific Sharpe Ratio'] =\
+        sharpe_ratio(specific_returns)
+
+    summary['Cumulative Specific Return'] =\
+        cum_returns_final(specific_returns)
+    summary['Cumulative Common Return'] =\
+        cum_returns_final(common_returns)
+    summary['Total Returns'] =\
+        cum_returns_final(total_returns)
+
+    summary = pd.Series(summary, name='')
+
+    annualized_returns_by_factor = [annual_return(perf_attrib[c])
+                                    for c in risk_exposures.columns]
+    cumulative_returns_by_factor = [cum_returns_final(perf_attrib[c])
+                                    for c in risk_exposures.columns]
+
+    risk_exposure_summary = pd.DataFrame(
+        data=OrderedDict([
+            (
+                'Average Risk Factor Exposure',
+                risk_exposures.mean(axis='rows')
+            ),
+            ('Annualized Return', annualized_returns_by_factor),
+            ('Cumulative Return', cumulative_returns_by_factor),
+        ]),
+        index=risk_exposures.columns,
+    )
+
+    return summary, risk_exposure_summary
+
+
+def _align_and_warn(returns,
```

Comment:
I moved this since it includes logic to make the indexes match up in addition to the warnings stuff. Should I make it so the warnings only get displayed if it's called from pyfolio?

Reply:
That's what I was thinking.  I see the warnings as being part of the tearsheet UX, not as part of the calculations themselves.